test/api-tests.js

```diff
@@ -218,6 +220,122 @@ describe('esprima.parse', function () {
 
 });
 
+describe('esprima.parse delegate', function () {
+
+    it('should receive all the nodes', function () {
+        var list = [];
+        function collect(node) {
+            list.push(node.type);
+        }
+        esprima.parse('/* universe */ answer = 42', {}, collect);
+
+        assert.deepEqual(list.length, 5);
+        assert.deepEqual(list, ['Identifier', 'Literal', 'AssignmentExpression', 'ExpressionStatement', 'Program']);
+    });
+
+    it('should receive the comments if specified', function () {
+        var list = [];
+        function collect(node) {
+            list.push(node);
+        }
+        esprima.parse('/* prolog */ answer = 42 // epilog', { comment: true }, collect);
+
+        assert.deepEqual(list.length, 7);
+        assert.deepEqual(list[0], { type: 'BlockComment', value: ' prolog ' });
+        assert.deepEqual(list[1], { type: 'Identifier', name: 'answer' });
+        assert.deepEqual(list[2], { type: 'LineComment', value: ' epilog' });
+        assert.deepEqual(list[3], { type: 'Literal', value: 42, raw: '42' });
+        assert.deepEqual(list[4].type, 'AssignmentExpression');
+        assert.deepEqual(list[5].type, 'ExpressionStatement');
+        assert.deepEqual(list[6].type, 'Program');
+    });
+
+    it('should be able to walk the tree and pick a node', function () {
+        var constant = null;
+        function walk(node) {
+            if (node.type === 'Literal') {
+                constant = node;
+            }
+        }
+        esprima.parse('answer = 42 // universe', {}, walk);
+        assert.deepEqual(constant, { type: 'Literal', value: 42, raw: '42' });
+    });
+
+    it('should be able to mutate each node', function () {
+        var ast = esprima.parse('42', { range: true }, function (node) {
+            node.start = node.range[0];
+            node.end = node.range[1];
+        });
+        assert.deepEqual(ast.start, 0);
+        assert.deepEqual(ast.end, 2);
+    });
+
+    it('should be affected by cover grammars', function () {
+        var list = [];
+        esprima.parse('(x, y) => z', {}, function (n) {
+            list.push(n.type)
+        });
+
+        // (x, y) will be a SequenceExpression first before being reinterpreted as
+        // the formal parameter list for an ArrowFunctionExpression
```

Comment:
Depending on the use case. If I use the delegate to perform syntax highlighting, it should be pretty harmless since the later arrow function expression overrides the "covered" sequence expression. If however I use the delegate for a static analyzer that prohibits any use of sequence expressions, then in some cases I need to delay the decision accordingly.

In all cases, if the callback logic can not handle with this sequence of events (pun intended!), then it is not suitable for an in-flight processing and there is no choice but to reimplement it as a post-parsing tree traversal.